core/opendaq/opendaq/src/instance_builder_impl.cpp

```diff
@@ -0,0 +1,311 @@
+#include <coreobjects/errors.h>
+#include <opendaq/instance_builder_impl.h>
+#include <opendaq/instance_builder_ptr.h>
+#include <opendaq/instance_ptr.h>
+#include <utility>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::GetOptions()
+{
+    return Dict<IString, IBaseObject>({
+        {"ModuleManager", Dict<IString, IBaseObject>({
+                {"ModulesPath", ""}
+            })},
+        {"Scheduler", Dict<IString, IBaseObject>({
+                {"ForceSingleThread", 0}
+            })},
+        {"Logging", Dict<IString, IBaseObject>()},
+        {"Modules", Dict<IString, IBaseObject>()}
+    });
+}
+
+InstanceBuilderImpl::InstanceBuilderImpl()
+    : componentsLogLevel(Dict<IString, LogLevel>())
+    , options(GetOptions())
+{
+}
+
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getModuleManagerOptions()
+{
+    return options.get("ModuleManager");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getSchedulerOptions()
+{
+    return options.get("Scheduler");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getLoggingOptions()
+{
+    return options.get("Logging");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getModuleOptions(IString* module)
+{
+    DictPtr<IString, IBaseObject> modules = options.get("Modules").asPtr<IDict>();
+    if (!modules.hasKey(module)) 
+    {   
+        auto moduleOptions = Dict<IString, IBaseObject>();
+        modules[module] = moduleOptions;
+        return moduleOptions;
+    }
+    return modules[module];
+}
+
+ErrCode InstanceBuilderImpl::build(IInstance** instance)
+{
+    if (instance == nullptr)
+        return OPENDAQ_ERR_ARGUMENT_NULL;
+
+    // Configire logger
+    if (!this->logger.assigned()) 
+    {
+        if (sinks.empty())
+            this->logger = Logger();
+        else
+        {
+            auto sinksList = List<ILoggerSink>();
+            for (const auto& sink : sinks)
+                sinksList.pushBack(sink);
+            
+            this->logger = LoggerWithSinks(sinksList);
+        }
+    }
+    
+    if (getLoggingOptions().hasKey("GlobalLogLevel"))
+        this->logger.setLevel(LogLevel(getLoggingOptions()["GlobalLogLevel"]));
+
+    for (const auto& [component, logLevel] : componentsLogLevel)
+    {
+        auto createdComponent = this->logger.getOrAddComponent(component);
+        createdComponent.setLevel(logLevel);
+    }
+
+    // Configure scheduler
+    if (!this->scheduler.assigned())
+        this->scheduler = Scheduler(this->logger, getSchedulerOptions()["ForceSingleThread"]);
+
+    // Configure moduleManager
+    if (!this->moduleManager.assigned())
+        this->moduleManager = ModuleManager(getModuleManagerOptions()["ModulesPath"]);
+
+    const auto builderPtr = this->borrowPtr<InstanceBuilderPtr>();
+    return daqTry([&]()
+    {
+        *instance = InstanceFromBuilder(builderPtr).detach();
+        return OPENDAQ_SUCCESS;
+    });
+}
```

Comment:
The entire building logic should be in the Instance constructor. This would allow a custom implementation of a builder to be provided. The `build` method should only call the appropriate factory.